refactor!: do not generate data for hidden columns

ugur-vaadin opened the pull request:
## Description

Value providers, tooltip generators, and part name generators run for every row regardless of column visibility. This adds unnecessary load when these callbacks do expensive work.

Each column's data generators are now wrapped in a `CompositeDataGenerator` that only runs them while the column is visible, and the row-level tooltip and part name generation are skipped for hidden columns too. Showing a previously hidden column refreshes the viewport so its data is generated again (but not reloaded from the data provider).

Depends on
- https://github.com/vaadin/flow/pull/24895
- https://github.com/vaadin/flow-components/pull/9661

> [!WARNING]
> Hidden columns no longer send data to the client side.

Fixes #6737

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
When adding a column, it sends two full updates to the client. One based on the old column config, and then another one with the new config. Also calls the data provider twice.

![Bildschirmfoto 2024-11-08 um 10 34 10](https://github.com/user-attachments/assets/376dee71-303e-45fc-a54d-58b8dfa714fc)

ugur-vaadin replied:
Added a fix and the test for it to handle this case.

ugur-vaadin commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -1181,15 +1155,54 @@ protected Column<?> getBottomLevelColumn() {
+                    .map(dataGenerator -> grid
+                            .addDataGenerator((DataGenerator) dataGenerator))
+                    .orElse(null);
+            grid.getDataProvider().refreshAll();
```
I agree that it would be better to do it that way. However, directly calling in `setVisible` does not work and I'm not sure what would be a good way to refactor the refresh.

vursen replied:
This is now resolved by calling `refreshViewport` instead of `refreshAll`. If both end up scheduled, DataCommunicator's scheduling system guarantees that only `refreshAll` runs, so nothing happens twice.

tomivirkki commented:
Would making the data generator run conditionally based on the column's visibility fix the original issue?
Proto: https://github.com/vaadin/flow-components/commit/1a5cb527e733f45369a2d08057847f4a53efa1cc

ugur-vaadin commented:
> Would making the data generator run conditionally based on the column's visibility fix the original issue? Proto: [1a5cb52](https://github.com/vaadin/flow-components/commit/1a5cb527e733f45369a2d08057847f4a53efa1cc)

Adapted the suggestion. It works nicely and is cleaner.

tomivirkki commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
In theory, hiding and unhiding columns shouldn't need to result in a data provider call, but only the data generators should need to be run for the already loaded items.

vursen replied:
Probably related: https://github.com/vaadin/flow-components/issues/6713

ugur-vaadin replied:
Added API to make this optional, and kept the default behaviour the same.

vursen replied:
I removed the opt-in API and enabled the new behavior by default, since with the new approach, toggling column visibility no longer causes any data provider requests.

vursen commented:
DataCommunicator needs a fix so it doesn't trigger DataProvider on `refreshViewport`. HierarchicalDataCommunicator already behaves this way.

I've added the corresponding tests to this PR, and one of them is currently failing because of that issue:

https://github.com/vaadin/flow-components/blob/19790aa5e069db5b9dde084202df4f64196925b8/vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridHiddenColumnRenderingTest.java#L159

https://github.com/vaadin/flow-components/blob/19790aa5e069db5b9dde084202df4f64196925b8/vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridHiddenColumnRenderingTest.java#L171

vursen committed:
feat!: skip data generation for hidden grid columns

Value providers, tooltip generators, and part name generators run for
every row regardless of column visibility, which adds unnecessary load
when these callbacks do expensive work.

Wrap each column's data generators in a CompositeDataGenerator that
only runs them while the column is visible, and skip the row-level
tooltip and part name generation for hidden columns too. Showing a
previously hidden column refreshes the viewport so its data is
generated again.

BREAKING CHANGE: hidden columns no longer send data to the client.
Depends on a matching fix in Flow so that refreshing the viewport
doesn't reload data from the data provider.

vursen committed:
remove unnecessary casting

vursen committed:
remove no longer needed setVisible override

vursen committed:
trigger CI

vursen committed:
test: rename columnTooltipGeneratorSpy to columnTooltipGeneratorCallCount

vursen committed:
trigger CI

sissbruecker commented on vaadin-grid-pro-flow-parent/vaadin-grid-pro-flow/src/main/java/com/vaadin/flow/component/gridpro/GridPro.java:
`GridPro` still generates cell editable data for hidden columns when a cell editable provider is set.

vursen replied:
Thanks, fixed.

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
There is an edge-case where `setVisible` runs in `DataCommunicator.flush` (e.g. reacting to `ItemCountChangeEvent`). In that case `grid.refreshViewport` doesn't have any effect and it sends empty cells.

I'd consider leaving this as is, but would like to hear your opinion.

vursen replied:
Good question. I'm actually inclined to call this a DataCommunicator bug: the `flushInProgress` flag should probably already be reset by the time events like ItemCountChangeEvent are fired, so that a  flush can be scheduled again. It's worth noting that the same issue can be reproduced with `refreshAll` when calling it from an ItemCountChangeEvent listener, so it's not only specific to `refreshViewport`.

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
There is an existing bug where toggling a column group's visibility forces columns to be visible on the client side due to how the web component works. Previously those columns would have data, now they may not.

I would be fine with filing an issue for this for now given that it's an existing problem. Another issue is that hiding columns on the server via the column group does not prevent sending data, though that could be a future improvement.

vursen replied:
Good catch! Filed a separate issue for this: https://github.com/vaadin/flow-components/issues/9677

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -449,6 +443,16 @@ public static class Column<T> extends AbstractColumn<Column<T>> {
 
+        private final CompositeDataGenerator<T> compositeDataGenerator = new CompositeDataGenerator<>() {
+            @Override
+            public void generateData(T item, ObjectNode jsonObject) {
```
There is also a `refreshData` which is called through `refreshItem`. So when refreshing items individually it will still run component renderers at least.

vursen replied:
Done: added the check and covered it with a test.

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -4367,6 +4338,9 @@ private void generateTooltipTextData(T item, ObjectNode jsonObject) {
         }
 
         idToColumnMap.forEach((id, column) -> {
+            if (!column.isVisible()) {
```
There would be an opportunity to register column tooltip / part name generators in the composite one to reuse the visibility check there. Generating the per row and per column tooltips would then need to look something like this:
```java
  private void generateTooltipTextData(T item, ObjectNode jsonObject) {                                                                                                                                                                                                                                                     
      String rowTooltip = tooltipGenerator.apply(item);                                                                                                                                                                                                                                                                     
      if (rowTooltip != null) {                                                                                                                                                                                                                                                                                             
          getOrCreateObject(jsonObject, "gridtooltips").put("row", rowTooltip);                                                                                                                                                                                                                                             
      }                                                                                                                                                                                                                                                                                                                     
  }

  compositeDataGenerator.addDataGenerator((item, jsonObject) -> {                                                                                                                                                                                                                                                           
      String cellTooltip = tooltipGenerator.apply(item);                                                                                                                                                                                                                                                                    
      if (cellTooltip != null) {                                                                                                                                                                                                                                                                                            
          getOrCreateObject(jsonObject, "gridtooltips")                                                                                                                                                                                                                                                                     
                  .put(getInternalId(), cellTooltip);                                                                                                                                                                                                                                                                       
      }                                                                                                                                                                                                                                                                                                                     
  });
```

Not sure if that's better, as it duplicates the property access then.

vursen replied:
I like this approach. Let's just do it in a follow-up PR.

vursen replied:
- https://github.com/vaadin/flow-components/pull/9679
- https://github.com/vaadin/flow-components/pull/9680

vursen committed:
fix: skip data refresh for hidden grid columns

Renderer's refreshData was still invoked for hidden columns since only
generateData was guarded by visibility, causing unnecessary rendering
work for columns not shown on the client.

vursen committed:
fix: skip cellEditableProvider for hidden grid pro columns

vursen committed:
format code
